### ISSUE TITLE
Surface reflection treatment: extract & live.or.die differ at high inclination

### ISSUE THREAD

**Issue by jhmatthews** [sirocco-rt/sirocco#524]
**Note: title updated. Originally a question based on misunderstanding but possibly still an issue to do with differences between extract and live or die.** 

I'm not yet sure if this is a problem but I think it's worth flagging. My understanding was that if I selected the "reflect" option for surface illumination treatment, then photons were perfectly reflected from the surface. If this is the case, then for an illustrative example with no wind, an isotropic SED with a reflecting disc (but no disc photons) should produce the same result as an isotropic SED without any disc surface at all, because a reflection is equivalent to travelling straight on into the opposite hemisphere (right?). 

Running this test reveals a flux difference in the case with the reflecting disc:
![nowind1](https://user-images.githubusercontent.com/3007249/59111273-9dcad380-8938-11e9-9553-9f4058cfc405.png)

whereas without a disc, all angles look identical:
![nowind](https://user-images.githubusercontent.com/3007249/59111271-9c011000-8938-11e9-821b-084dab2dd158.png)

To run this test, use this pf file and change disc type to "flat" or "none". 
[nowind.pf.txt](https://github.com/agnwinds/python/files/3266316/nowind.pf.txt)

**Comment by kslong** [comment]
Sounds like a problem.  With a reflecting disk the photon actually goes straight through I believe.

**Comment by jhmatthews** [comment]
The problem goes away when I switch to live or die, so this appears to be a problem about extracting photons that hit surfaces. 
EDIT: this was a mistake, see comment below.

**Comment by kslong** [comment]
That sounds possible to me.  I don't remember ever thinking about that part of the problem, either for this case, or for the case where the photon is supposed to scatter at that point.

**Comment by jhmatthews** [comment]
It looks to me like the issue might be in these lines of trans_phot.c (in this case line 373):

```C
      if (geo.absorb_reflect == BACK_RAD_SCATTER)
      {
        /* If we got here, the a new photon direction needs to be defined that will cause the photon
         * to continue in the wind.  Since this is effectively a scattering event we also have to
         * extract a photon to consturct the detailed spectrum
         */
        randvcos (pp.lmn, normal);
        stuff_phot (&pp, p);
        tau_scat = -log (1. - random_number (0.0, 1.0));
        istat = pp.istat = P_INWIND;
        tau = 0;
        if (iextract)
        {
          stuff_phot (&pp, &pextract);
          extract (w, &pextract, PTYPE_DISK);
        }
      }
```

I don't think that's doing the right thing and it looks like it's introducing a costheta random direction, which could cause the above behaviour.

**Comment by kslong** [comment]
I had forgotten about this change , which were introduced to try to handle #260  where we changed from simply reflection to scattering.  But this looks correct to me. 

The first part of the code you show is just the normal photon going through the wind.  But if this is in the spectral cycles, we also have to treat this as a scattering site, and so we extract a photon as well.  I think one needs to look at what extract does at this point.  Is the photon not getting blocked  correctly.  

(At any rate, we should try to close out #260 at the same time as we solve this.

**Comment by jhmatthews** [comment]
I don't really understand Knox - doesn't the call to randvcos above introduce an anisotropic direction, appropriate for a disk emission, but not the same as a reflection and with no knowledge of the incident angle? 

It's not really clear to me why extract has to be called here either. Can't we just let the photon carry on? A true reflection is the same as carrying straight on, and we don't extract when we cross e.g. a cell boundary, do we?

**Comment by kslong** [comment]
We changed what reflection meant from mirror reflection to scattering, so this is why randvcos is needed for the photons that are going through the winds.  We did not keep  mirror reflection as a separate option.

With this change, you will, I think, get different answers for a system with a star and a disk compared to a star without a disk, because the radiation with a disk will end up being anisotropic (more directed to the poles).  The total luminosity of the system though should stay the same, which is something that should be checked.

For the purposes of extracting a detailed spectrum, reflection is a scattering event, which is why extract is called when it encounter disk.  Now normally extract would encounter the disk for photons that are on the “back” side.  What I am not sure about, and probably should be checked,  is whether this is always effective when the photon hits the disk surface.    I cannot immediately think of a good way to check this, except by following photons in detailed.

**Comment by jhmatthews** [comment]
Ok - so we want to introduce an anisotropic distribution, I guess I'm fine with that. 
I realised that my statement about live or die above is wrong. Live or die does have anisotropy, but it's not quite the same as the extract case (note the difference in flux to the extract spectrum above)
![nowind_liveordie](https://user-images.githubusercontent.com/3007249/59503180-a28f0a80-8e97-11e9-834a-5c070bf7b7ff.png)

**Comment by kslong** [comment]
@jhmatthews   Can we close this issue.  The original premise that reflection meant effective specular reflection is no longer what we do.  Do we want to implement specular reflection, as a separate option.

**Comment by jhmatthews** [comment]
I think the original issue was solved, in that it was due to a misunderstanding on my part, but I think there is still a possible difference between live or die and extract here (as indicated by my most recent comment).

**Comment by jhmatthews** [comment]
I retested this. Updated pf file atttached. Extract and live or die basically produce the same results from a reflecting disc, except at very high inclinations (89 degrees) there are some differences in flux. Here's a comparison for an AGN model with mdot_wind set to 1e-20 (i.e. transparent) and a reflecting disc:
 
![ext-v-lod](https://user-images.githubusercontent.com/3007249/81092909-37a85100-8ef9-11ea-8a35-56e1bbb45014.png)

I think the difference at 89 degrees  is because live or die uses DANG_LIVE_OR_DIE to decide whether a photon contributes to an angular bin, and the bin is defined in degrees space. Thus, at the equator, 2 degrees occupies more solid angle than elsewhere, so the flux is overestimated. So there is no issue here, except that if we want live or die to be a functioning mode we should probably define DANG_LIVE_OR_DIE as a delta_costheta rather than a delta_theta. Other than that issue can be closed. Any thoughts @kslong or others?

[nowind.pf.txt](https://github.com/agnwinds/python/files/4582322/nowind.pf.txt)

**Comment by kslong** [comment]
I am not entirely sure. While it is true that the angles are described in degree space, there is a renormalization factor that is applied, see line 203 of spectra.c where it is calculated, that seems to take the cos theta term out.  I am fairly suspicious that this is a problem with the boundaries going through 90 degrees.  It looks like if you took a angle of 90 degrees, the renormalization would be infinitely large.  In particular, I don't understand these two lines

````
x1 = fabs (cos (x1 / RADIAN));
x2 = fabs (cos (x2 / RADIAN));
````

and in particular the fabs bit which are subtracted to the dominator of the renormalization factor.  It looks like to me that if you took 90 degrees for the angle, one might get 0 for the denominatory

**Comment by jhmatthews** [comment]
Yep I agree. Specifically: when you choose 90 degrees, then x1 and x2 are equal because you take fabs(cos(92)) and fabs(cos(88)) - as the DANG is 2 degrees. If you choose 89 degrees, then you underestimate the difference between the cosines because one of them is negative, but the fabs makes it positive. This means the normalisation factor is ~a factor 2 higher than it should be. 

However, that still doesn't seem to be the full story, because even with the fabs() removed, there are still differences between extract and live or die. Here's the ratio of the two methods at 89 degrees with and without the fabs: 

![liveordie_ratio](https://user-images.githubusercontent.com/3007249/81157663-1be78e00-8f7f-11ea-8aae-fac7836555f4.png)

**Comment by mandyhewitt** [comment]
@jhmatthews @kslong I may have recently experienced this issue in my polarization analytical model at angles near 90 degrees. I was trying to work out why my results were showing twice the flux expected for the highest angle.  I will check my .pf and model spectra results and update here. My analytical model had no disk and I was using extract at 89.9 degrees.

**Comment by mandyhewitt** [comment]
@jhmatthews @kslong I had a look at my analytical model results again today to remind myself of the issue which I was having last year and how the flux was eventually numerically calculated correctly (and then corresponded with python output flux). The error in my calculations was because I was using the two radii of the wind as listed in the imported model parameter file. This meant that I was using the central radius of the wind instead of the outer radius of the wind. This meant that my flux calculations were half of the observed flux in the spectrum. I'm not sure if this is related in any way here but thought I should mention it.

**Comment by jhmatthews** [comment]
This is very strange. I've tried a few runs with a CV model with and without a radiating disk and reflecting disk, and with and without the fabs() term in the above code. 

**with the fabs term** the models without a radiating disk (cv_nowind and cv_nodisk) look fine even if they have a reflecting disk. The model with a radiating disk (cv_diskrad) has an enhanced flux in live.or.die mode as above with the original AGN type model. Here's a plot comparing live or die and extract in the 3 models.

![comparison_fabs](https://user-images.githubusercontent.com/3007249/82542139-62291980-9b49-11ea-8a43-a0b4e7abfcf8.png)

**without the fabs term** the models without a radiating disk now look wrong and the live.or.die flux is below the extract flux, so I must be missing something. The model with a radiating disk has an enhanced flux in live.or.die mode but it's only marginally enhanced. Here's a plot comparing live or die and extract in the 3 models.

![comparison_nofabs](https://user-images.githubusercontent.com/3007249/82542296-a4525b00-9b49-11ea-9f80-e9c232092262.png)

Finally here's a table summarising:
|  Run type | radiation sources   |  reflecting disk? | live.or.die matches extract w/ fabs term |  live.or.die matches extract with no fabs term |
|---|---|---|---|---|
|  CV     |  WD+disk  | yes  |  no (l.or.d > extract)  | nearly (l.or.d>extract) | 
|  CV     |  WD only   | yes  |  yes | no (l.or.d < extract) | 
|  CV     |  WD only   | no   |  yes  | no (l.or.d < extract)   |

**Comment by kslong** [comment]
@jhmatthews  I am having a little trouble duplicating this.  Here is my plot using today's dev version of cv_reflect in the two modes.  I have only modified what is in the regression test by changing some of the angles to be close to the disk.   Can you provide .pf files to illustrate what you mean.

![cv_reflect](https://user-images.githubusercontent.com/2530901/82698476-71cf5d80-9c30-11ea-8b21-784f63f4f1cf.png)

**Comment by jhmatthews** [comment]
woops, I meant to upload it. The basic model is just the CV model but with the wind mass loss rate  set to a small number. Then, this particular model has a reflecting disk and a radiating disk, and is an extract model. You can fiddle with disk radiation (yes/no), disk type (none/flat) and extract/live.or.die to get my above results. 

[cv_diskrad.pf.txt](https://github.com/agnwinds/python/files/4687975/cv_diskrad.pf.txt)

**Comment by kslong** [comment]
This issue needs to be investigated, but today I closed issue #834 which sounds a bit similar.  In that case, the final problem was not longer about surface reflection, but really about differences in live or die and extract when computing spectra along the wind cone.  This I think is different.
